pkg/linter/linter.go

```diff
@@ -512,105 +492,39 @@ func pythonTestPostLinter(_ LinterContext, fsys fs.FS) error {
 	return nil
 }
 
-// Checks if the linters in the given slice are known linters
-// Returns an empty slice if all linters are known, otherwise a slice with all the bad linters
-func CheckValidLinters(check []string) []string {
-	linters := []string{}
-	for _, l := range check {
-		_, present := linterMap[l]
-		if present {
-			continue
-		}
-
-		// Check post-linter map too
-		_, present = postLinterMap[l]
-		if !present {
-			linters = append(linters, l)
-		}
-	}
-
-	return linters
-}
-
-func (lctx LinterContext) lintPackageFs(warn func(error), linters []string, linterClass linter_defaults.LinterClass) error {
+func lintPackageFs(pkgname string, fsys fs.FS, linters []string) error {
 	// If this is a compat package, do nothing.
-	if strings.HasSuffix(lctx.pkgname, "-compat") {
+	if strings.HasSuffix(pkgname, "-compat") {
 		return nil
 	}
 
-	// Verify all linters are known
-	badLints := CheckValidLinters(linters)
-	if len(badLints) > 0 {
-		return fmt.Errorf("unknown linter(s): %s", strings.Join(badLints, ", "))
-	}
-
-	postLinters := []string{}
-	walkCb := func(path string, d fs.DirEntry, err error) error {
-		if err != nil {
-			return fmt.Errorf("error traversing tree at %s: %w", path, err)
+	errs := []error{}
+	for _, linterName := range linters {
+		linter, found := linterMap[linterName]
+		if !found {
+			return fmt.Errorf("unknown linter: %q", linterName)
 		}
-
-		for _, linterName := range linters {
-			linter, present := linterMap[linterName]
-			if !present {
-				// We already checked that all linters are valid, so this must be a post linter
-				postLinters = append(postLinters, linterName)
-				continue
-			}
-
-			if linter.LinterClass&linterClass == 0 {
-				// Linter not in class, ignored
-				continue
-			}
-
-			err = linter.LinterFunc(lctx, path, d)
-			if err != nil {
-				if linter.FailOnError {
-					return fmt.Errorf("linter %s failed at path %q: %w; suggest: %s", linterName, path, err, linter.Explain)
-				}
-				warn(err)
-			}
+		if err := linter.LinterFunc(linterName, fsys); err != nil {
+			errs = append(errs, fmt.Errorf("linter %s failed; suggest: %s", linterName, linter.Explain))
 		}
-
-		return nil
 	}
 
-	if err := fs.WalkDir(lctx.fsys, ".", walkCb); err != nil {
-		return err
-	}
-
-	// Run post-walking linters
-	for _, linterName := range postLinters {
-		linter := postLinterMap[linterName]
-
-		if linter.LinterClass&linterClass == 0 {
-			// Linter not in class, ignored
-			continue
-		}
-
-		err := linter.LinterFunc(lctx, lctx.fsys)
-		if err != nil {
-			if linter.FailOnError {
-				return fmt.Errorf("linter %s failed; suggest: %s", linterName, linter.Explain)
-			}
-			warn(err)
-		}
-	}
-
-	return nil
+	return errors.Join(errs...)
 }
 
 // Lint the given build directory at the given path
-func LintBuild(packageName string, path string, warn func(error), linters []string) error {
+func LintBuild(ctx context.Context, packageName string, path string, require, warn []string) error {
```

Comment:
WDYT about moving the warn vs require distinction outside of this function?

Then we'd just call it twice like:

```go
if err := linter.LintBuild(ctx, pkg, path, warns); err != nil {
    log.Warnf("package linter warning in %s: %v", packageName, err)
}
if err := linter.LintBuild(ctx, pkg, path, requires); err != nil {
  return err
}
```

Or is it better for the require vs warn handling to be the responsibility of this function/package?

Genuinely asking, I haven't thought too deeply about this.

Reply:
I don't really love this either. I don't feel strongly, it's only done twice.